### ISSUE TITLE
Set-BuildEnvironment: Some concerns about environment variables

### ISSUE THREAD

**Issue by nightroman** [RamblingCookieMonster/BuildHelpers#10]
I have some concerns about environment variables. They will do in many cases,
indeed. Besides, unlike PowerShell variables, they are available for invoked
applications. The choice of environment variables is understandable.

### Limitations

At the same time for pure PowerShell scenarios environment variables are not
natural. While "not natural" is an academic argument, they are potentially
troublesome in some practical cases, typically with more than one project
involved.

### Nested builds

There may be projects consisting of several nested projects. For example,
see *FarNet family* at [Build-Scripts-in-Projects](https://github.com/nightroman/Invoke-Build/wiki/Build-Scripts-in-Projects).
Calling `Set-BuildEnvironment` for a child build is going to alter the
previously set environment of the current build.

### Parallel builds

Several projects may be built in parallel runspaces (e.g. by `Invoke-Builds.ps1` of `Invoke-Build`).
Parallel runspaces still share the same process environment. So use of
`Set-BuildEnvironment` in its current form is probably not possible.

### Suggestion

Can we have an extra command similar to `Set-BuildEnvironment` which sets
PowerShell variables with some standardized names rather than environment
variables? Or, to avoid similar commands, let `Set-BuildEnvironment` do
this optionally? Or even by default and set environment optionally, when
this is really needed and with mentioned limitations of scenarios.

**Comment by RamblingCookieMonster** [comment]
Hi Roman!

Makes sense to me - not sure where and how this module is used, would it makes sense to add a validateset parameter (no idea on the name: `-By`, `-Medium`, `-As`, etc., which takes one or more strings, and defaults to `ENV` or `EnvironmentVariable`, but allows for `Variable`?

On that note, would it be sensible to add a `Prefix` that defaults to BH, allowing folks to specify alternatives?

Cheers!

**Comment by nightroman** [comment]
Thank you for understanding the concerns, this is important. The "how" part is up to you. Interested folks may have ideas and speak, too. I do not have much. (`-As Variable` (parameter + argument), `-AsVariable` (switch), etc.).

BTW, if I am not mistaken, it is not straightforward to set variables in the parent scope from a script module function. But this should be possible, one way or another.

`BH` looks clear to me. If somebody asks for an optional custom prefix then of course.

**Comment by nightroman** [comment]
Just one thought. Perhaps the switch `-AsVariable` is better because it may
open a new parameter set with more parameters like `Scope`. A typical use case.
`Set-BuildEnvironment` is called from init-like functions or tasks and they
probably want to set variables not in their scopes but either in a parent scope
or, often, at the build script scope, so that then other functions or tasks use
the build variables.

P.S. The project is an excellent idea. Build helpers in PowerShell may be
related to tools like git and MSBuild but they do not have to be bound to
any task or build runner.

**Comment by RamblingCookieMonster** [comment]
Thanks for the advice!

If folks have any better ideas for implementing this, open to suggestions.  As is, mapping a Scope parameter to the Set-Variable Scope parameter.

**Comment by nightroman** [comment]
I took a quick look, it seems to be working. But I will test it a little bit more.

Would it be better if we use the *Script* scope by default? *Global* is a bad practice, as such. Besides, it is not suitable for nested builds, and the original concern was exactly about nested builds.

**Comment by RamblingCookieMonster** [comment]
Sure, not sure why I went with global, does seem an odd choice : )  Will swap it out before deploying

**Comment by nightroman** [comment]
Then this is done. Thank you!

**Comment by nightroman** [comment]
It is not straightforward... We are probably hitting an issue similar to [Not-found-variable](https://github.com/nightroman/PowerShellTraps/tree/master/Module/Not-found-variable).
The command is working in simple scenarios and not working in slightly more complex.

Here is the problem case. Two scripts in the same directory and this directory is the current location.

The script `Test-1.1.ps1` works (with some expected errors to be discussed later, perhaps):

````powershell
    Import-Module ...\BuildHelpers\BuildHelpers\BuildHelpers.psd1

    $ErrorActionPreference = 'Continue'
    Set-BuildEnvironment -AsVariable -Scope Script

    Get-Variable BH* -Scope Script
    'done'
````

Yet another script `Test-1.2.ps1` calls the first. No variables, as a result!

````powershell
    .\Test-1.1.ps1
````

**Comment by RamblingCookieMonster** [comment]
I had to resort to scope numbers.  For example:

test-1.1.ps1:

```powershell
Import-Module C:\users\wframe\documents\github\BuildHelpers\BuildHelpers -force

$ErrorActionPreference = 'Continue'
Set-BuildEnvironment -AsVariable -Scope 1 #-Path C:\temp\PSDeploy

Get-Variable BH* -Scope 1
'done'
```

The above code worked when calling .\test-1.1.ps1 from the appropriate path (or specifying the path)

Using test-1.2.ps1 with same content as yours does not work for this.  It does work if I make a change to test-1.1.ps1:

```powershell
Get-Variable BH* -Scope 2 # get parent of parent's scope
```

This is all far above my head though.  Not sure I'll be much help.

Cheers!

**Comment by nightroman** [comment]
Unless there is a better idea, I would use the following solution.

The module exports an alias (so a different command name) `Set-BuildVariable`.
In .psm1 this alias is resolved to the script `Set-BuildVariable.ps1`
(the script, not dot-sourced function!):

    Set-Alias Set-BuildVariable "$PSScriptRoot\..\Set-BuildVariable.ps1"

The script `Set-BuildVariable.ps1` sets variables in the specified scope.
Unlike the similar module function, the script should have no issues.

This is a fair work around known script module issues.

**Comment by RamblingCookieMonster** [comment]
Interesting idea! Sounds good to me, feel free to submit a pull request if you're up for it, otherwise will try to put it together tonight or over the next few days.

Cheers!

**Comment by nightroman** [comment]
Yes, I can do this (in a few hours, in the morning). If you would like me to do this could you please revert last changes in `Set-BuildEnvironment`? Then I will just add a new script and an alias. Should I create a new directory for the script? Perhaps there will be more than one script eventually, for similar or other reasons. What is the name then, *Scripts*?

**Comment by RamblingCookieMonster** [comment]
Awesome!  Yep!  A Scripts folder would work

**Comment by nightroman** [comment]
My fork is [here](https://github.com/nightroman/BuildHelpers) with the alias
and script added. It works fine as discussed, i.e. the known issues are
resolved. But there is a new issue, so I am not proposing a PR yet.

The scope for now defaults to 1, i.e. the caller scope:

````powershell
    # defaults to -Scope 1, the caller scope
    Set-BuildVariable -Path $PSScriptRoot
    Get-Variable BH* -Scope 0
````

Note that the default is not *Script*. Why? The following command works

````powershell
     Set-BuildVariable -Path $PSScriptRoot -Scope Script
````

but it creates variables in the *Script* scope of *Set-BuildVariable.ps1*, of course...
This is correct but useless for a caller that wants to create variables in its script scope.

I do not know how to create variables in a script scope which is the parent of
the currently invoked script, at least without ugly hacks.

It is still possible to achieve this legally by dot-sourcing the command:

     . Set-BuildVariable -Path $PSScriptRoot -Scope Script

Now the code is invoked in the calling scope and *Script* works for its script
scope, as required.

It does not look neat perhaps but it will do if we do not have better options
and document this properly. Let me know what you think. If you agree with this
approach then I have some more work to do:

- If the command is not dot-sourced and the specified scope is *Script* then an
  error should be thrown, otherwise variables are created "nowhere" for a caller.
- Used local variables of *Set-BuildVariable.ps1* should be reviewed (renamed)
  for avoiding conflicts with variables in the calling scope. This is not an
  option for parameters `$Path` and `$Scope`, they still may conflict.

****

A side note about naming: I realized that some commands use nouns in plural form:
`Get-BuildVariables.ps1`, `Get-ModuleFunctions.ps1`, `Set-ModuleFunctions.ps1`.
This is not recommended by PowerShell naming guidelines. But if this is by
design then the new command `Set-BuildVariable` should follow it, too.

**Comment by RamblingCookieMonster** [comment]
Hi!

Ahh, fun with scopes : )

Your call on invocation - if it would be helpful to allow calling it via `. Set-BuildVariable`, I don't see any issues (outside of a more interesting implementation).

On function naming: I know.  I'm bad.  I suppose it would be good to follow best practices.  I'll rename the functions and (for now) add aliases to the old pluralized names.

**Comment by nightroman** [comment]
Very well, I will complete this and send the PR later today.
